docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+
+  It will not try to fetch data from obj store, and return an error if some pages in the range have not been fetched.
+
+  `Range` internals:
+
+  - convert range to page ids  
+  - if all page ids have been fetched
+    - return a slice of mmaped file
+
+  * otherwise return an error
+
+* CloseMmap: `CloseMmap` unmmaps `f.data` .
+
+* CloseFile: `CloseFile` closes  `f.file`. 
+
+* Close: `Close` closes `f.file`, unmmaps `f.data` and set `f.pages` to nil.
+
+### Range to pages
+
+Convert a range to pages:
+
+```
+const pageSize = 4 * 1024  // 4KiB
+
+minPageID := math.Floor( float(start) / float(pageSize) )
+maxPageID := math.Ceil( float(end) / float(pageSize) )
+
+pages := roaring.NewBitmap()
+pages.AddRange(minPageID, maxPageID)
+```
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next  time to fetch. So that we can combine multiple queries to ones.
+
+When fetching combined  pages,
+if an error occurs, BucketFile will send the error to callers via the `pendingChan` channel,
+if no error occurs, BucketFile will close the `pendingChan` channel.
+
+If callers receive an error from the channel, then prefetch failed. 
+If callers receive nil (zero value of error, after channel closed) from the channel, then prefetch done. 
+
+Each time to fetch pending pages，BucketFile will reset  `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory object
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into memory object for sequential and  random accesses.  
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses very small memory. 
+
+For inactive IndexReaders, if the inactive duration (`now.Sub(lastQueryTime)`) greats than a configured duration (e.g. 15 mins), then IndexReader will remove in-memory objects and close file descriptor and mmap in IndexFile.  This will reduce memory usage.
+
+**NOTE**: The `pages` bitmap in BucketFile should always be kept in memory until the server is closed.
+
+### Preload blocks on startup
+
+The latest blocks will be queried mostly. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local Files
+
+Because the max size of a index file is 64 GiB, and  the max size of a chunks file is 128 MiB.
```

Comment:
Really? I think chunks files are capped on 0.5GB

Reply:
you are right. it should be 512MiB.
128MiB is the size of WAL file.

Reply:
updated.